File: lockSeq.py
P = {2, 3, 5, 7, 11, 13, 17}
mod = 10**9 + 7


def f(i, last):
    ans = dp[i][last]
    if ans != -1:
        return ans % mod
    ans = 0
    if i == N:
        ans = 1
    else:
        for dig in xrange(10):
            if i == 0:
                ans += f(i + 1, dig) % mod
            elif last + dig not in P:
                ans += f(i + 1, dig) % mod
    dp[i][last] = ans
    return ans

if __name__ == '__main__':
    N = input()
    dp = [[-1] * 11 for i in xrange(100005)]
    print f(0, 0) % mod

File: code-cpp/lockSeq.cpp
#include <bits/stdc++.h>
using namespace std;

typedef  long long ll;
ll N;

#define mod 1000000007

vector<ll> P = {0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0};
ll dp[100005][10];

ll f(ll i, ll last){
    ll ans = dp[i][last];
    if (ans != -1)
        return ans;
    ans = 0;
    if (i == N)
        ans = 1;
    else{
        for (ll dig = 0; dig < 10; dig++){
            if (i == 0)
                ans = (ans + f(i + 1, dig)) % mod;
            else if (!P[dig + last])
                ans = (ans + f(i + 1, dig)) % mod;
        }
    }
    dp[i][last] = ans % mod;
    return ans;
}



int main(){
    memset(dp, -1, sizeof(dp));
    cin >> N;
    cout << f(0, 0) << endl;

    return 0;
}
